### ISSUE TITLE
[Task 28] 实现 RLOO advantage estimator

### ISSUE THREAD

**Issue by xcosmosbox** [redai-infra/Relax#186]
我认领 Task 28。
目标是在算法 registry 中实现 RLOO 的 baseline、advantage、专用 loss 与监控指标，只覆盖同步（colocate）训练。

### 现状

任务的起点是 `redai-infra/Relax` main@`8a54679e971087566bfda939a5f75649e07fb861`，相关结构如下：

- 算法经 `--advantage-estimator` 分发，现存七种 estimator。
- GRPO 家族的组内奖励变换集中在 `relax/utils/utils.py::post_process_rewards`：按 `Sample.group_index` 分组、mean-centering、可选除以组内 std。这个调用发生在 `convert_samples_to_train_data` 内，早于任何 DP 切分。
- token 级只做标量广播（`ppo_utils.py::get_grpo_returns`）。
- loss 层已经有 `compute_sapo_loss`/`compute_cispo_loss`，专用 loss 的扩展模式是现成的。

RLOO 需要对同组 G 个采样（G = `--n-samples-per-prompt`），

```text
baseline_i = ( Σ_{j≠i} r_j ) / (G-1)
A_i        = r_i - baseline_i = (G/(G-1)) · (r_i - r̄)      # 实现用缩放式，单测双向验证
L_i,t      = -sg(A_i) · log π_θ(y_i,t)                      # 非裁剪 REINFORCE，逐有效 token
```

和 GRPO 的差别：不除组内 std。RLOO 的基线与样本 i 独立，梯度无偏，同时保留了奖励尺度。和 `reinforce_plus_plus_baseline` 的差别：不强制全局 whitening、advantage 里不折 KL、loss 非裁剪。

### 改动计划

7 个文件，尽量保证最小 diff，并使得他们的默认行为不变：

1. `relax/utils/training/ppo_utils.py`：新增 `compute_rloo_leave_one_out_rewards(group_rewards)`，计算 `(G/(G−1))·(r−mean)`。这个函数只接受 1-D 输入——传进 2-D 会被整体当成一个超大组归约，静默算错，所以必须防御性拒绝。另加 G≥2 断言和 `torch.isfinite` fail-fast。再新增 `compute_rloo_loss(log_probs, advantages)`：逐元素 `-(A.detach()·logπ)`，clipfrac 恒 0，与 sapo/cispo 同构。模块顶层不引入 megatron。
2. `relax/utils/utils.py::post_process_rewards`：分组分支加入 `"rloo"`，按 estimator 分流调用 LOO 变换；GRPO 族逻辑逐字不动。`get_debug_data` 的组感知子采样列表要同步扩展，否则调试子采样会把组切碎、破坏基线。
3. `relax/utils/arguments.py`：choices 加 `"rloo"`；新增六条启动校验（见下表）。
4. `relax/core/registry.py`：`ALGOS["rloo"]` 用与 `"grpo"` 相同的角色映射（无 critic）。
5. `relax/backends/megatron/loss.py`：标量广播分发表追加 `"rloo"`；`policy_loss_function` 在 sapo/cispo 分发处新增一个自包含的 `elif rloo` 分支，调用 `compute_rloo_loss`。
6. `relax/components/advantages.py`：广播分发表同步追加，同步/异步两张表保持一致（异步已被校验条条拦截）。
7. `relax/utils/metrics/metric_utils.py`：RLOO 诊断指标（见下）。

**启动校验**（arguments.py，六条都在服务启动前给出可读报错和修复建议）：

| 拒绝 | 原因 |
|---|---|
| `--fully-async`（含 hybrid-async） | 本题仅同步；异步路径未验证，拒绝而非静默放行 |
| `--n-samples-per-prompt < 2` | G−1=0，LOO 基线不存在 |
| `--disable-rewards-normalization` | 会跳过构建基线的组处理路径，静默退化为无基线 REINFORCE |
| `--normalize-advantages` | 该开关在 DP 切分后做 `distributed_masked_whiten`：既重新引入 RLOO 刻意去除的 std 归一化，又使 advantage 依赖切分方式、破坏 DP 不变性 |
| `rollout_batch_size × n_samples_per_prompt ≠ global_batch_size` | 非裁剪目标无 ratio 校正，每 rollout 恰一次 optimizer 更新是正确性前提；违反时第二次更新是无校正的 off-policy 训练 |
| `--partial-rollout` / `--use-dynamic-global-batch-size` | 每步有效样本数漂移会破坏上一条前提；v1 未验证路径不静默放行 |

**方案完成目标**：

- **loss**：专用非裁剪 REINFORCE（上式），不复用 PPO-Clip。理由有两层：RLOO 是因为无偏才被选中的，对它施加信任域在动机上自相矛盾；而且复用裁剪目标得到的其实是"带 LOO 基线的 GRPO"，不是论文的 RLOO。实现走仓库既有的专用 loss 扩展模式，数学上等价于 CISPO 把重要性系数固定为 1。连带后果也一并冻结：`--eps-clip/--eps-clip-high` 对 rloo 失效、`train/pg_clipfrac` 恒 0（正好用作实跑接线自检）；loss 归约沿用框架 reducer（recipe 两臂同用 `--calculate-per-token-loss`），论文形式的序列级求和开关不纳入 v1——CP 下要先跨 rank gather 完整 logπ，成本不对称。`pg_loss` 均值可以为负（`-A·logπ`，logπ<0），这是预期行为，稳定性看 `entropy_loss` 与 `grad_norm`。
- **KL**：与本仓库 GRPO 相同，loss 侧可选 `--use-kl-loss --kl-loss-coef`；reward 侧 KL 折入不实现。
- **边界行为**：上表六条校验；组大小不符沿用既有 ValueError；NaN/±Inf reward fail-fast，报错带 group/样本位置，并有测试断言 GRPO 行为不变；空 response 的 reward 照常参与他人基线、自身零 loss 零梯度（如果丢弃它，会偏置同组其他样本的基线）；组内奖励全相等时 advantage 全 0，靠指标监控。
- **DP/CP 口径**：DP 不变性在构造上成立——LOO 在 `post_process_rewards` 完成，那里是 DP 切分之前的全量 batch，组不会跨 DP 破碎；前置条件正是上面那条 normalize-advantages 校验。CP 是标量广播 + CP-aware 归约。验证 = 构造论证 + CPU 恒等式单测 + 真实 `torch.distributed` gloo 进程组测试（world_size 2/4 × per-token-loss on/off，各 rank 部分归约之和等于 cp_size=1 的结果，且切片恰好把每条 response 划分一次，排除靠抵消凑出来的相等；无需 GPU）。验证边界要说清楚：我的自备硬件是 1×A100-40GB，跑不了真实多卡 DP/CP 对照，如需多卡实测请指出。
- **监控指标**：`rloo/baseline_mean` 挂在 rollout 侧——raw_reward 和变换后的 rewards 只在那里同时可见，基线可以直接输出；loss 侧微批不携带 raw_reward。`rloo/adv_abs_mean`：组内 advantage 和恒为 0，普通均值没有信息量，所以用绝对值衡量信号强度。`rloo/no_signal_frac`：token 级零优势占比，零优势 token 不贡献梯度，主因是组内奖励全相等，算是奖励饱和的前哨信号；按 token 而不是按组统计，因为 G≥3 时单个样本也可能恰好落在组均值上。另有 `rloo/empty_response_frac`、`rloo/zero_adv_group_frac`，加上既有的 `rollout/raw_reward`、`train/pg_loss、pg_clipfrac、ppo_kl、entropy_loss、grad_norm` 等。
- `arguments.py` 属于仓库 AGENTS.md 的 Ask-First 范围：六条校验以本 Issue 作为事前确认。

### 单测计划

- LOO 数学：暴力 `Σ_{j≠i}/(G−1)` 逐元素对照（float64，atol=1e-12）；缩放恒等式（G∈{2,3,5,8,16}）；全相等组；G=2；非 1-D 输入拒绝。
- `post_process_rewards`：乱序 group_index 分组；raw_rewards 不变；组大小错误/非有限 reward 报错，且 GRPO 行为不受影响。
- loss 逐元素：`pg_loss == -(A·logπ)`（含混合长度与 padding）；clipfrac 恒 0；autograd 验证 `d pg_loss/d logπ == -A`；负向测试：取会触发裁剪的 ratio，断言与 PPO-Clip 结果不同。
- DP/CP：全批 LOO 后切分片恒等；广播切片恒等；gloo 真实进程组 CP 归约测试（如上面的方案完成目标所示）。
- 配置面：choices 接受 rloo；六条校验逐一触发；默认 grpo 路径回归不变；registry 角色映射与 grpo 一致。

### 实验安排

固定环境：

- Branch起点：main@`8a54679e971087566bfda939a5f75649e07fb861`，从该 commit 叉出
- 运行时： 固定环境，与任务仓库对齐
- 硬件：单卡与双卡或八卡×NVIDIA A100-PCIE-40GB（colocate 单卡分时）
- 模型：`Qwen/Qwen3-0.6B`；数据：GSM8K train 7473 题，`--rm-type math`

运行矩阵与口径：两臂共用同一脚本和全部参数，只切换 `ADVANTAGE_ESTIMATOR`。`ROLLOUT_BATCH_SIZE=16`、`N_SAMPLES=8`（G=8）、`GLOBAL_BATCH_SIZE=128`（满足等式约束条件，每 rollout 恰 1 次更新）、`NUM_ROLLOUT=60`、response 上限 2048、temperature 1、lr 1e-6 constant、两臂均不开 KL loss、两臂统一 `--clip-grad 0`。clip-grad 关掉的原因要说清楚：GRPO 除以组内 std、RLOO 乘 G/(G−1)，两臂产生的优势尺度不同，梯度范数分布也跟着不同，默认 clip-grad 1.0 会以不同频率触发两臂的梯度裁剪，等于把裁剪差异混进算法对比；稳定性以 grad_norm/entropy 护栏监控。其余：grpo 臂 `--eps-clip 0.2`（rloo 无此参数）、`--use-rollout-logprobs`。主对比 A1=GRPO(seed 42) → B1=RLOO(seed 42)；时间允许补 B2=RLOO(43)、A2=GRPO(43) 成 ABBA。

在线校验（贯穿全程）：rloo 臂 `train/pg_clipfrac` 恒 0（接线自检）；每个 rollout 查 `rollout/advantages` 组内和 ≈ 0（1e-7 量级）；`entropy_loss` 无塌缩。

复现命令模板：

```bash
ADVANTAGE_ESTIMATOR=rloo NUM_ROLLOUT=60 ROLLOUT_BATCH_SIZE=16 N_SAMPLES=8 GLOBAL_BATCH_SIZE=128 \
MODEL_DIR=<models目录> DATA_DIR=<datasets目录> \
bash examples/algorithms/run-qwen3-0.6B-1xgpu-rloo.sh --rollout-seed 42 \
  --tb-experiment-name task28-b1-rloo-s42
# 对照臂仅改 ADVANTAGE_ESTIMATOR=grpo 与实验名
```

记录指标：`rollout/raw_reward`、`rollout/rewards`、`rollout/advantages`、`rollout/response_lengths`、`passrate/*`、`train/loss、pg_loss、pg_clipfrac、ppo_kl、entropy_loss、grad_norm`、`rloo/*` 诊断；吞吐（samples/s、response tokens/s、rollout/train 阶段耗时）、平均 GPU 利用率与峰值显存（nvidia-smi 旁路采样）。

正确性护栏（任一违反则该臂作废重跑）：全程无 NaN/Inf、无 OOM、每步 128 样本齐全、grad_norm 有限、无异常退出与静默降级、结束后 Ray/SGLang/GPU 进程清理干净。

结果判读口径：本题是算法基线补全，不是性能优化。主交付是相同预算下 RLOO vs GRPO 的 reward / loss / KL（`train/ppo_kl`）/ 吞吐对比曲线和三窗口数值表；吞吐差异预期 <2%，是护栏不是收益指标；不预设 RLOO 优于 GRPO，差异、方差、标准误都如实报告。

### 交付方式

提交 PR、单元测试、实验报告、文档。
